pkg/lib/clusterconfig/clusterconfig.go

```diff
@@ -260,33 +298,168 @@ var UserValidation = &cr.StructValidation{
+			if err != nil {
+				return errors.Wrap(err, InstanceDistributionKey)
+			}
+
+			compatibleInstanceCount++
+		}
+
+		if compatibleInstanceCount == 0 {
+			return ErrorAtLeastOneInstanceDistribution(chosenInstanceType, compatibleSpots[0].Type)
+		}
+
+		if cc.OnDemandBaseCapacity != nil && *cc.OnDemandBaseCapacity > *cc.MaxInstances {
+			return ErrorOnDemandBaseCapacityGreaterThanMax(*cc.OnDemandBaseCapacity, *cc.MaxInstances)
+		}
+	} else {
+		if len(cc.InstanceDistribution) > 0 {
+			return ErrorConfiguredWhenSpotIsNotEnabled(InstanceDistributionKey)
+		}
+		if cc.OnDemandBaseCapacity != nil {
+			return ErrorConfiguredWhenSpotIsNotEnabled(OnDemandBaseCapacityKey)
+		}
+		if cc.OnDemandPercentageAboveBaseCapacity != nil {
+			return ErrorConfiguredWhenSpotIsNotEnabled(OnDemandPercentageAboveBaseCapacityKey)
+		}
+		if cc.MaxPrice != nil {
+			return ErrorConfiguredWhenSpotIsNotEnabled(InstanceDistributionKey)
+		}
+		if cc.SpotInstancePools != nil {
+			return ErrorConfiguredWhenSpotIsNotEnabled(SpotInstancePoolsKey)
+		}
+	}
+
+	return nil
 }
 
-func PromptValidation(skipPopulatedFields bool, promptInstanceType bool, defaults *ClusterConfig) *cr.PromptValidation {
+func CheckCortexSupport(instanceMetadata aws.InstanceMetadata) error {
+	if strings.HasSuffix(instanceMetadata.Type, "nano") ||
+		strings.HasSuffix(instanceMetadata.Type, "micro") ||
+		strings.HasSuffix(instanceMetadata.Type, "small") {
+		ErrorInstanceTypeTooSmall()
+	}
+	if instanceMetadata.GPU > 0 && !strings.HasPrefix(instanceMetadata.Type, "p2") && !strings.HasPrefix(instanceMetadata.Type, "p3") {
+		return ErrorGPUInstanceTypeNotSupported(instanceMetadata.Type)
+	}
+	return nil
+}
+
+func CheckSpotInstanceCompatibility(target aws.InstanceMetadata, suggested aws.InstanceMetadata) error {
+	if target.GPU > suggested.GPU {
+		return ErrorIncompatibleSpotInstanceTypeGPU(target, suggested)
+	}
+
+	if target.Memory.Cmp(suggested.Memory) > 0 {
+		return ErrorIncompatibleSpotInstanceTypeMemory(target, suggested)
+	}
+
+	if target.CPU.Cmp(suggested.CPU) > 0 {
+		return ErrorIncompatibleSpotInstanceTypeCPU(target, suggested)
+	}
+
+	return nil
+}
+
+func CompatibleSpotInstances(targetInstance aws.InstanceMetadata) ([]aws.InstanceMetadata, error) {
+	compatibleInstances := []aws.InstanceMetadata{}
+	instanceMap := aws.InstanceMetadatas[targetInstance.Region]
+	for instanceType, instanceMetadata := range instanceMap {
+		if instanceType == targetInstance.Type {
+			continue
+		}
+
+		if err := CheckCortexSupport(instanceMetadata); err != nil {
+			continue
+		}
+
+		if err := CheckSpotInstanceCompatibility(targetInstance, instanceMetadata); err != nil {
+			continue
+		}
+
+		compatibleInstances = append(compatibleInstances, instanceMetadata)
+	}
+	if len(compatibleInstances) == 0 {
+		return compatibleInstances, ErrorNoCompatibleSpotInstanceFound(targetInstance.Type)
+	}
+	return compatibleInstances, nil
+}
+
+func (cc *ClusterConfig) AutoFill() error {
```

Comment:
If this only auto-fills spot-related fields, we can call this something like `AutoFillSpot()`, and put this at the top:

```go
if cc.Spot == nil || !*cc.Spot {
    return nil
}
```